#Team Members and IDs:
**Michael Danial** 13004528, **Ali Tharwat** 13004222, **Sherif Tamer** 13004065


# We Couldnt use the same dataset as Milestone 1

Due to the fact that the **Loan Approval** dataset does not include the loan approval decision

according to the dataset provider :
https://www.kaggle.com/datasets/suryadeepthi/loan-approval-dataset/data

as its written in  **About Dataset** section :
This dataset provides a realistic simulation of loan applications, containing 5000 entries and 15 features that reflect key factors influencing loan approvals. It is designed for exploratory data analysis (EDA), feature engineering, and machine learning modeling.

The dataset **does not include the loan approval decision**, allowing users to build their own prediction models.

# We used the **Employee Attrition Prediction Dataset** Instead :
https://www.kaggle.com/datasets/ziya07/employee-attrition-prediction-dataset/data

# 1. Data Preparation :

In [ ]:
import pandas as pd

# Load dataset
df = pd.read_csv('/content/employee_attrition_dataset.csv')
print("Initial Dataset Shape:", df.shape) # Number of Rows & Number of Columns

Initial Dataset Shape: (1000, 26)


In [ ]:
#a) Categorical Encoding:

from sklearn.preprocessing import LabelEncoder

# Identify categorical columns
categorical_cols = df.select_dtypes(include=['object']).columns
print('Categorical Columns : ', categorical_cols.tolist())

# Apply Label Encoding to all categorical columns
for col in categorical_cols: df[col] = LabelEncoder().fit_transform(df[col])

Categorical Columns :  ['Gender', 'Marital_Status', 'Department', 'Job_Role', 'Overtime', 'Attrition']


In [ ]:
#b) Split the Data :

# Set 'y' as the target () and drop it from X
X = df.drop('Attrition', axis=1)
y = df['Attrition']

In [ ]:
#c) Training and Testing Sets:

from sklearn.model_selection import train_test_split

# Split data into 75% training and 25% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

# 2. Apply Machine Learning Algorithms:


In [ ]:
#a) K-Nearest Neighbors (KNN): K = 5

from sklearn.neighbors import KNeighborsClassifier

# Initialize KNN model with 5 neighbors
knn = KNeighborsClassifier(n_neighbors=5)

# Train the model using the training data
knn.fit(X_train, y_train)

# Predict the target values for the test set
y_pred_knn = knn.predict(X_test)

In [ ]:
#b) Naive Bayes:

from sklearn.naive_bayes import GaussianNB

# Initialize the Naive Bayes model
nb = GaussianNB()

# Train the model on the training data
nb.fit(X_train, y_train)

# Predict the target values for the test set
y_pred_nb = nb.predict(X_test)

In [ ]:
#c) Additional Models :
#  I) Decision Tree , II) Support Vector Machine (SVM) , III) Random Forest


# I) Decision Tree:
from sklearn.tree import DecisionTreeClassifier

# Initialize the Decision Tree model
dt = DecisionTreeClassifier(random_state=42)

# Train the model using the training data
dt.fit(X_train, y_train)

# Predict the target values for the test set
y_pred_dt = dt.predict(X_test)

#---------------------------------------------------------------

# II) Support Vector Machine (SVM) :
from sklearn.svm import SVC

# Initialize the SVM model (using default RBF kernel)
svm = SVC(random_state=42)

# Train the model on the training data
svm.fit(X_train, y_train)

# Predict the target values for the test set
y_pred_svm = svm.predict(X_test)

#---------------------------------------------------------------

# III) Random Forest :
from sklearn.ensemble import RandomForestClassifier

# Initialize the Random Forest model
rf = RandomForestClassifier(random_state=42)

# Train the model using the training data
rf.fit(X_train, y_train)

# Predict the target values for the test set
y_pred_rf = rf.predict(X_test)

In [ ]:
#d)Bonus: Deep Learning Model :

#  Feedforward Neural Network/Multilayer Perceptron (MLP).

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from sklearn.preprocessing import StandardScaler

# Scale features for better neural network performance
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Build the model
model = Sequential([
    Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    Dense(32, activation='relu'),
    Dense(1, activation='sigmoid')  # Output layer for binary classification
])

# Compile the model
model.compile(optimizer=Adam(learning_rate=0.001), loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
history = model.fit(X_train_scaled, y_train, epochs=20, batch_size=32, verbose=1, validation_split=0.2)

# Predict the target values for the test set
y_pred_dl_prob = model.predict(X_test_scaled)
y_pred_dl = (y_pred_dl_prob > 0.5).astype(int).flatten()

Epoch 1/20


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


19/19 ━━━━━━━━━━━━━━━━━━━━ 2s 28ms/step - accuracy: 0.3569 - loss: 0.8270 - val_accuracy: 0.7600 - val_loss: 0.5675
Epoch 2/20
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.7653 - loss: 0.5600 - val_accuracy: 0.8000 - val_loss: 0.5060
Epoch 3/20
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.8060 - loss: 0.5011 - val_accuracy: 0.8000 - val_loss: 0.5051
Epoch 4/20
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8011 - loss: 0.4834 - val_accuracy: 0.8000 - val_loss: 0.5066
Epoch 5/20
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8012 - loss: 0.4865 - val_accuracy: 0.8000 - val_loss: 0.5076
Epoch 6/20
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7836 - loss: 0.4858 - val_accuracy: 0.8000 - val_loss: 0.5095
Epoch 7/20
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8030 - loss: 0.4631 - val_accuracy: 0.8000 - val_loss: 0.5109
Epoch 8/20
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.7832 - loss: 0.4780 - val_accuracy: 0.8000 - val_loss: 0.5132


# 3. Model evaluation:

#A) for each model we will evaluate:

*   Accuracy
*   Confusion Matrix
*   Precision
*   Recall

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score

In [ ]:
# KNN :
print("K-Nearest Neighbors (KNN) Evaluation:")

#     Accuracy
accuracy_knn = accuracy_score(y_test, y_pred_knn)
print("Accuracy:", accuracy_knn)

#     Confusion Matrix
conf_matrix_knn = confusion_matrix(y_test, y_pred_knn)
print("Confusion Matrix:\n", conf_matrix_knn)

#     Precision
precision_knn = precision_score(y_test, y_pred_knn)
print("Precision:", precision_knn)

#     Recall
recall_knn = recall_score(y_test, y_pred_knn)
print("Recall:", recall_knn)

K-Nearest Neighbors (KNN) Evaluation:
Accuracy: 0.816
Confusion Matrix:
 [[200  12]
 [ 34   4]]
Precision: 0.25
Recall: 0.10526315789473684


In [ ]:
# Naive Bayes :
print("Naive Bayes Evaluation:")

#     Accuracy
accuracy_nb = accuracy_score(y_test, y_pred_nb)
print("Accuracy:", accuracy_nb)

#     Confusion Matrix
conf_matrix_nb = confusion_matrix(y_test, y_pred_nb)
print("Confusion Matrix:\n", conf_matrix_nb)

#     Precision
precision_nb = precision_score(y_test, y_pred_nb)
print("Precision:", precision_nb)

#     Recall
recall_nb = recall_score(y_test, y_pred_nb)
print("Recall:", recall_nb)

Naive Bayes Evaluation:
Accuracy: 0.848
Confusion Matrix:
 [[212   0]
 [ 38   0]]
Precision: 0.0
Recall: 0.0


/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [ ]:
# Decision Tree :
print("Decision Tree Evaluation:")

#     Accuracy
accuracy_dt = accuracy_score(y_test, y_pred_dt)
print("Accuracy:", accuracy_dt)

#     Confusion Matrix
conf_matrix_dt = confusion_matrix(y_test, y_pred_dt)
print("Confusion Matrix:\n", conf_matrix_dt)

#     Precision
precision_dt = precision_score(y_test, y_pred_dt)
print("Precision:", precision_dt)

#     Recall
recall_dt = recall_score(y_test, y_pred_dt)
print("Recall:", recall_dt)

Decision Tree Evaluation:
Accuracy: 0.68
Confusion Matrix:
 [[162  50]
 [ 30   8]]
Precision: 0.13793103448275862
Recall: 0.21052631578947367


In [ ]:
# Support Vector Machine (SVM) :
print("Support Vector Machine (SVM) Evaluation:")

#     Accuracy
accuracy_svm = accuracy_score(y_test, y_pred_svm)
print("Accuracy:", accuracy_svm)

#     Confusion Matrix
conf_matrix_svm = confusion_matrix(y_test, y_pred_svm)
print("Confusion Matrix:\n", conf_matrix_svm)

#     Precision
precision_svm = precision_score(y_test, y_pred_svm)
print("Precision:", precision_svm)

#     Recall
recall_svm = recall_score(y_test, y_pred_svm)
print("Recall:", recall_svm)

Support Vector Machine (SVM) Evaluation:
Accuracy: 0.848
Confusion Matrix:
 [[212   0]
 [ 38   0]]
Precision: 0.0
Recall: 0.0


/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [ ]:
# Random Forest :
print("Random Forest Evaluation:")

#     Accuracy
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print("Accuracy:", accuracy_rf)

#     Confusion Matrix
conf_matrix_rf = confusion_matrix(y_test, y_pred_rf)
print("Confusion Matrix:\n", conf_matrix_rf)

#     Precision
precision_rf = precision_score(y_test, y_pred_rf)
print("Precision:", precision_rf)

#     Recall
recall_rf = recall_score(y_test, y_pred_rf)
print("Recall:", recall_rf)

Random Forest Evaluation:
Accuracy: 0.848
Confusion Matrix:
 [[212   0]
 [ 38   0]]
Precision: 0.0
Recall: 0.0


/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [ ]:
# Bonus : Deep Learning(MLP):
print("Deep Learning (MLP) Evaluation:")

#     Accuracy
accuracy_dl = accuracy_score(y_test, y_pred_dl)
print("Accuracy:", accuracy_dl)

#     Confusion Matrix
conf_matrix_dl = confusion_matrix(y_test, y_pred_dl)
print("Confusion Matrix:\n", conf_matrix_dl)

#     Precision
precision_dl = precision_score(y_test, y_pred_dl)
print("Precision:", precision_dl)

#     Recall
recall_dl = recall_score(y_test, y_pred_dl)
print("Recall:", recall_dl)

Deep Learning (MLP) Evaluation:
Accuracy: 0.848
Confusion Matrix:
 [[212   0]
 [ 38   0]]
Precision: 0.0
Recall: 0.0


/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


#B) Comparison :

**📌 Key Observations:**

* Naive Bayes, SVM, Random Forest, and MLP achieved higher accuracy (82+%) , However , they failed to predict any positive cases, resulting in 0 precision and recall. This suggests these models are simply predicting the majority class (0), likely due to class imbalance.


* Decision Tree attempted to classify both classes
 , had higher recall (21%), meaning it captured more true positives than any other model, even if accuracy(68%) was lower.

* KNN was the most balanced overall:
it predicted both classes, had a moderate accuracy (82%), and achieved precision (25%), which was the highest among all models — though its recall was low (~10%).

Best Performing Model:

**K-Nearest Neighbors (KNN)**

It’s the only model with a reasonable balance between predicting both classes and maintaining decent accuracy.

Despite others having slightly better accuracy, they completely fail at identifying positive cases — rendering them ineffective for real-world use where false negatives matter (e.g., predicting employee attrition).

KNN’s higher precision indicates that when it predicts a positive case, it’s more likely to be correct.